src/Toolchain.ml

```diff
@@ -0,0 +1,495 @@
+                    match kv with
+                    | Ok kv -> kv :: acc
+                    | Error msg ->
+                      Js.log msg;
+                      acc)
+                  []
+             |> Js.Dict.fromList |> R.return)
+
+  let setupToolChain workspaceRoot spec =
+    let { cmd; kind } = spec in
+    match kind with
+    | Global -> Ok () |> Js.Promise.resolve
+    | Esy root ->
+      let rootStr = root |> Fpath.toString in
+      Cmd.output cmd ~args:[| "status"; "-P"; rootStr |] ~cwd:rootStr
+      |> P.then_ (fun r ->
+             let r =
+               match r with
+               | Error _ -> R.return false
+               | Ok stdout -> (
+                 match Json.parse stdout with
+                 | None -> R.return false
+                 | Some json ->
+                   json
+                   |> (let open Json.Decode in
+                      field "isProjectReadyForDev" bool)
+                   |> R.return )
+             in
+             Js.Promise.resolve r)
+      |> P.then_ (function
+           | Error e -> e |> R.fail |> Js.Promise.resolve
+           | Ok isProjectReadyForDev ->
+             if isProjectReadyForDev then
+               () |> R.return |> Js.Promise.resolve
+             else if Fpath.compare root workspaceRoot = 0 then (
+               Js.log "esy project";
+               promptSetup ~f:(fun () ->
+                   Window.withProgress
+                     [%bs.obj
+                       { location = Window.locationToJs Window.Notification
+                       ; title = "Setting up toolchain..."
+                       }]
+                     (fun progress ->
+                       (progress.report
+                          [%bs.obj { increment = int_of_float 1. }] [@bs]);
+                       Cmd.output cmd ~cwd:(root |> Fpath.toString) ~args:[||]
+                       |> P.then_ (fun _ ->
+                              (progress.report
+                                 [%bs.obj { increment = int_of_float 100. }]
+                               [@bs]);
+                              Js.Promise.resolve (Ok ()))))
+             ) else (
+               Js.log "bsb project";
+               setupWithProgressIndicator cmd ~envWithUnzip:Process.env
+                 (root |> Fpath.toString)
+             ))
+    | Opam _ -> Js.Promise.resolve (Ok ())
+
+  let lsp spec =
+    let { cmd; kind } = spec in
+    match kind with
+    | Opam _ -> (Cmd.binPath cmd, [| "exec"; "ocamllsp" |])
+    | Esy root -> (Cmd.binPath cmd, [| "-P"; Fpath.toString root; "ocamllsp" |])
+    | Global -> ("ocamllsp", [||])
+
+  let compare x y =
+    match (x, y) with
+    | Esy root1, Esy root2 -> Fpath.compare root1 root2
+    | Opam root1, Opam root2 -> Fpath.compare root1 root2
+    | Global, Global -> 0
+    | Opam _, Esy _ -> -1
+    | Esy _, Opam _ -> 1
+    | Esy _, Global -> -1
+    | Opam _, Global -> -1
+    | Global, _ -> 1
+
+  let rec find name = function
```

Comment:
is `List.find_map` available in bucklescript?

Reply:
I don't think so. You can do something like that with `Belt.List.get` → `Belt.Option.map` or `Belt.List.keepMap` → `Belt.List.get`

Reply:
I see. We can use `Belt` or roll our own `find_map`.